python/rpdk/java/templates/init/guided_aws/TagHelper.java

```diff
@@ -0,0 +1,250 @@
+                Map.Entry::getValue));
+    }
+
+    /**
+     * getTagsToRemove
+     *
+     * Determines the tags the customer desired to remove from the function.
+     */
+    public Set<String> generateTagsToRemove(final Map<String, String> currentTags, final Map<String, String> desiredTags) {
+        final Set<String> desiredTagNames = desiredTags.keySet();
+
+        return currentTags.keySet().stream()
+            .filter(tagName -> !desiredTagNames.contains(tagName))
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * getTagsToAdd
+     *
+     * Determines the tags the customer desired to define or redefine.
+     */
+    public Set<Tag> generateTagsToAdd(final Set<Tag> currentTags, final Set<Tag> desiredTags) {
+        return Sets.difference(new HashSet<>(desiredTags), new HashSet<>(currentTags));
+    }
+
+    /**
+     * getTagsToRemove
+     *
+     * Determines the tags the customer desired to remove from the function.
+     */
+    public Set<Tag> generateTagsToRemove(final Set<Tag> currentTags, final Set<Tag> desiredTags) {
+        return Sets.difference(new HashSet<>(currentTags), new HashSet<>(desiredTags));
+    }
+
+    /**
+     * doUpdateTags
+     *
+     * This function updates the target resource's user defined tags. It first
+     * calculate what tags to add and remove, and then calls TagResource/UntagResource
+     * to achieve the customer's desired set of tags as specified in their CloudFormation template.
+     *
+     */
+    public ProgressEvent<ResourceModel, CallbackContext> doUpdateTags(final AmazonWebServicesClientProxy proxy, final ProxyClient<SdkClient> serviceClient,
+                                                                      final ResourceModel resourceModel, final ResourceHandlerRequest<ResourceModel> handlerRequest, final CallbackContext callbackContext, final Logger logger) {
+        return ProgressEvent.progress(resourceModel, callbackContext)
+            .then(progress -> {
+                final Map<String, String> tagsToAdd = generateTagsToAdd(getCurrentTags(handlerRequest), getNewDesiredTags(resourceModel, handlerRequest));
+                if (!tagsToAdd.isEmpty()) {
+                    return tagResource(proxy, serviceClient, resourceModel, handlerRequest, callbackContext, tagsToAdd, logger);
+                }
+                return ProgressEvent.defaultInProgressHandler(callbackContext, 0, resourceModel);
+            })
+            .then(progress -> {
+                final Set<String> tagsToRemove = generateTagsToRemove(getCurrentTags(handlerRequest), getNewDesiredTags(resourceModel, handlerRequest));
+                if (!tagsToRemove.isEmpty()) {
+                    return untagResource(proxy, serviceClient, resourceModel, handlerRequest, callbackContext, tagsToRemove, logger);
+                }
+                return ProgressEvent.defaultInProgressHandler(callbackContext, 0, resourceModel);
+            });
+    }
```

Comment:
do we really need it? I can see it being confusing

Reply:
Yeah I was thinking it can be the only method for service team to use inside updateHandler. By calling doUpdateTags every other things will be done inside our TagHelper class.